File: lib/fmt.lua
-- fmt.lua


local max, min, abs = math.max, math.min, math.abs
local floor, log = math.floor, math.log
local char, byte = string.char, string.byte


local function hsvANSI(h, s, v)
	if h ~= h then h = 0 end
	h = h * 6
	local r = (1 + s * (min(max(abs((h + 0) % 6 - 3) - 1, 0), 1) - 1)) * v
	local g = (1 + s * (min(max(abs((h + 4) % 6 - 3) - 1, 0), 1) - 1)) * v
	local b = (1 + s * (min(max(abs((h + 2) % 6 - 3) - 1, 0), 1) - 1)) * v

	r, g, b = math.floor(r * 255), math.floor(g * 255), math.floor(b * 255)
	return ("\x1b[38;2;%d;%d;%dm"):format(r, g, b)
end


local function intPart(x)
	local b = 2 ^ floor(log(x, 2))
	local b1 = 2 ^ floor(log(x, 2) + 1)
	local ret = (1 / b1 + x % b / b)
	return ret
end

local nums = {}
local aliases = {}

local function _tonum(str)
	for i, n in pairs(nums) do
		if str == n then
			return "#"..i
		end
	end
	return false
end

local function _toalias(str)
	for s, v in pairs(aliases) do
		if str == s then
			return v
		end
	end
	return false
end

local function _str(expr, depth, alias)
	if not expr then return "NIL" end
	if depth > 20 then return "!!!" end
	if type(expr) == "table" then
		if #expr == 1 then
			local v = char(depth % 26 + 97)
			local a = _str(expr[1], depth + 1, alias)
			if a:match"^%b()$" then a = a:sub(2, -2) end
			local ret = ('(\xff%s.%s)'):format(v, a)
			ret = alias and _tonum(_str(expr, 0, false)) or ret
			ret = alias and _toalias(_str(expr, 0, false)) or ret
			return ret
		end
		local left = _str(expr[1], depth, alias)
		local right = _str(expr[2], depth, alias)
		left = alias and _tonum(_str(expr[1], 0, false)) or left
		right = alias and _tonum(_str(expr[2], 0, false)) or right
		left = alias and _toalias(_str(expr[1], 0, false)) or left
		right = alias and _toalias(_str(expr[2], 0, false)) or right
		return ("(%s%s)"):format(left, right)
	else
		return char((depth - expr) % 26 + 97)
	end
end

aliases[_str(T, 0, false)] = "T"
aliases[_str(F, 0, false)] = "F"
for i = 1, 100 do
	nums[i] = _str(N(i), 0, false)
end

local function colorLetter(letter)
	local value = byte(letter) - byte"a"
	return hsvANSI(intPart(value), .5, .9)..letter.."\x1b[0m"
end

local function colorNumber(num)
	return hsvANSI(intPart(tonumber(num:sub(2))), .5, .9)..num.."\x1b[0m"
end

local function colorAlias(alias)
	if alias == "T" then
		return hsvANSI(.33, .8, .9)..alias.."\x1b[0m"
	elseif alias == "F" then
		return hsvANSI(.00, .8, .9)..alias.."\x1b[0m"
	else
		return hsvANSI(0, 0, 1)..alias.."\x1b[0m"
	end
end

---@param expr table|number @expression to stringify
---@param lambda string|nil @prefix for abstractions
---@param color boolean|nil @whether to colorize the output
---@param group boolean|nil @whether to group abstraction variables
---@param alias boolean|nil @whether to detect church numerals
---@return string
local function stringify(expr, lambda, color, group, alias)
	lambda = lambda or "@"
	local str = _str(expr, 0, alias)
	if str:match"^%b()$" then str = str:sub(2, -2) end
	while group and str ~= str:gsub("([a-z])%.\xff([a-z])", "%1%2") do
		-- group abstraction variables
		-- \x.\y.\z.e -> \xyz.e
		str = str:gsub("([a-z])%.\xff([a-z])", "%1%2")
	end
	str = str:gsub("\xff", lambda)
	if color then
		str = str:gsub("%l", colorLetter)
		str = str:gsub("#%d+", colorNumber)
		for _, pattern in pairs(aliases) do
			str = str:gsub(pattern, colorAlias)
		end
		return str
	end
	return str
end

return stringify

File: py/lib/fmt.py
# fmt.py

import math
from .defs import N, T, F


def hsv_ansi(h, s, v):
    h *= 6

    def channel(offset):
        return (1 + s * (min(max(abs((h + offset) % 6 - 3) - 1, 0), 1) - 1)) * v

    r = math.floor(channel(0) * 255)
    g = math.floor(channel(4) * 255)
    b = math.floor(channel(2) * 255)
    return f"\x1b[38;2;{r};{g};{b}m"


def int_part(x):
    try:
        b = 2 ** math.floor(math.log(x, 2))
        b1 = 2 ** math.floor(math.log(x, 2) + 1)
    except ValueError:
        return 0
    return 1 / b1 + x % b / b


nums = {}
aliases = {}


def _tonum(string):
    for i, n in nums.items():
        if string == n:
            return f"#{i}"
    return False


def _toalias(string):
    for s, v in aliases.items():
        if string == s:
            return v
    return False


def _str(expr, depth=0, alias=True):
    if expr is None:
        return "NIL"
    if depth > 20:
        return "!!!"
    if isinstance(expr, list):
        if len(expr) == 1:
            v = chr(depth % 26 + 97)
            a = _str(expr[0], depth + 1, alias)
            ret = f"(\xff{v}.{a})"
            if alias:
                alt = _str(expr, 0, False)
                ret = _tonum(alt) or _toalias(alt) or ret
            return ret
        else:
            left = _str(expr[0], depth, alias)
            right = _str(expr[1], depth, alias)
            if alias:
                left_alt = _str(expr[0], 0, False)
                right_alt = _str(expr[1], 0, False)
                left = _tonum(left_alt) or _toalias(left_alt) or left
                right = _tonum(right_alt) or _toalias(right_alt) or right
            return f"({left}{right})"
    else:
        return chr((depth - expr) % 26 + 97)


aliases[_str(T, 0, False)] = "T"
aliases[_str(F, 0, False)] = "F"
for i in range(1, 101):
    nums[i] = _str(N(i), 0, False)


def color_letter(letter):
    value = ord(letter) - ord("a")
    return f"{hsv_ansi(int_part(value), 0.5, 0.9)}{letter}\x1b[0m"


def color_number(num):
    value = int(num[1:])
    return f"{hsv_ansi(int_part(value), 0.5, 0.9)}{num}\x1b[0m"


def color_alias(alias):
    if alias == "T":
        return f"{hsv_ansi(0.33, 0.8, 0.9)}{alias}\x1b[0m"
    elif alias == "F":
        return f"{hsv_ansi(0.00, 0.8, 0.9)}{alias}\x1b[0m"
    else:
        return f"{hsv_ansi(0, 0, 1)}{alias}\x1b[0m"


def stringify(expr: list | int, lambda_char="λ", color=False, group=False, alias=False):
    string = _str(expr, 0, alias)
    if group:
        import re

        pattern = r"([a-z])\.\xff([a-z])"
        while re.search(pattern, string):
            string = re.sub(pattern, r"\1\2", string)
    string = string.replace("\xff", lambda_char)
    if color:
        import re

        string = re.sub(r"[a-z]", lambda m: color_letter(m.group(0)), string)
        string = re.sub(r"#\d+", lambda m: color_number(m.group(0)), string)
        for pattern in aliases.values():
            string = string.replace(pattern, color_alias(pattern))
        return string
    return string
